syntax = "proto3";

package products;

// Product Service Definition
service ProductService {
    rpc CreateProduct(CreateProductRequest) returns (ProductResponse);
    rpc GetProduct(GetProductRequest) returns (ProductResponse);
    rpc ListProducts(ListProductsRequest) returns (ListProductsResponse);
    rpc UpdateProduct(UpdateProductRequest) returns (ProductResponse);
    rpc DeleteProduct(DeleteProductRequest) returns (DeleteProductResponse);
    rpc SearchProducts(SearchProductsRequest) returns (ListProductsResponse);
}

// Messages
message Product {
    int32 id = 1;
    string name = 2;
    string description = 3;
    double price = 4;
    int32 stock_quantity = 5;
    string category = 6;
    string created_at = 7;
    string updated_at = 8;
}

message CreateProductRequest {
    string name = 1;
    string description = 2;
    double price = 3;
    int32 stock_quantity = 4;
    string category = 5;
}

message GetProductRequest {
    int32 id = 1;
}

message ListProductsRequest {
    int32 page = 1;
    int32 page_size = 2;
}

message ListProductsResponse {
    repeated Product products = 1;
    int32 total_count = 2;
    int32 page = 3;
    int32 page_size = 4;
}

message UpdateProductRequest {
    int32 id = 1;
    string name = 2;
    string description = 3;
    double price = 4;
    int32 stock_quantity = 5;
    string category = 6;
}

message DeleteProductRequest {
    int32 id = 1;
}

message DeleteProductResponse {
    bool success = 1;
    string message = 2;
}

message SearchProductsRequest {
    string query = 1;
    string category = 2;
    double min_price = 3;
    double max_price = 4;
}

message ProductResponse {
    Product product = 1;
    bool success = 2;
    string message = 3;
}
